sql/sql_acl.cc

```diff
@@ -14442,7 +14450,8 @@ bool acl_authenticate(THD *thd, uint com_change_user_pkt_len)
     case CR_AUTH_USER_CREDENTIALS:
       errors.m_authentication= 1;
       if (thd->password && !mpvio.make_it_fail)
-        handle_password_errors(acl_user->user.str, acl_user->host.hostname, PASSWD_ERROR_INCREMENT);
+        handle_password_errors(acl_user->user.str, acl_user->host.hostname,
```

Comment:
Though was a problem with the previous implementation, we've got a `acl_user` (`const` needs dropping),  and we're taking its user/host to look the user up again in `handle_password_errors`? Seems wasteful. Can `handle_password_errors` and `update_password_errors` be dropped and a `clear_password_errors` / `increment_password_errors` exist in the `AC_USER` class.

I'm assuming the `acl_cache->lock` was related to the `find_user_exact` function, but if its for the increment/clear itself, perhaps `password_errors` should be a `Atomic_counter`.

Reply:
acl_cache->lock guards all acl_users and some other related data structures. One must hold the lock anyway for the small period of time need to find and increment the counter. In the future we could improve this, but not in the scope of this bug fix.

Thank you for looking into this.